hubble-relay: use distroless as the base image and run as non-root

rolinh opened the pull request:
In order to improve the security posture of Hubble Relay, this patch updates the base image for Hubble Relay from scratch to distroless. Distroless images come in different variants and the one being used here is the most basic one that only contains the following:

  - ca-certificates
  - A /etc/passwd entry for a root, nonroot and nobody users
  - A /tmp directory
  - tzdata

Given that this new base image comes with CA certificates, we no longer need to import CA certificates from the Alpine image. Moreover, the hack for running gops, namely setting `ENV GOPS_CONFIG_DIR=/` is no longer required. Finally, the patch sets the image user to the nonroot user with UID 65532.

At last, to run as non-root, the securityContext for the Hubble Relay container is updated to drop all capabilities and run as the user:group 65532:65532.


```release-note
Run Hubble Relay as non-root user by default.
```

rolinh committed:
hubble-relay: use distroless as the base image and run as non-root

In order to improve the security posture of Hubble Relay, this patch
updates the base image for Hubble Relay from scratch to distroless.
Distroless images come in different variants and the one being used here
is the most basic one that only contains the following:

  - ca-certificates
  - A /etc/passwd entry for a root, nonroot and nobody users
  - A /tmp directory
  - tzdata

Given that this new base image comes with CA certificates, we no longer
need to import CA certificates from the Alpine image. Moreover, the hack
for running gops, namely setting `ENV GOPS_CONFIG_DIR=/` is no longer
required. Finally, the patch sets the image user to the nonroot user
with UID 65532.

At last, to run as non-root, the securityContext for the Hubble Relay
container is updated to drop all capabilities and run as the user:group
65532:65532.

Signed-off-by: Robin Hahling <robin.hahling@gw-computing.net>

michi-covalent reviewed (approved):
nice!

chancez reviewed (changes requested):
Might be worth adding a note in `upgrade.rst`.

qmonnet commented on images/hubble-relay/Dockerfile:
```diff
@@ -1,9 +1,17 @@
 # SPDX-License-Identifier: Apache-2.0
 
-ARG BASE_IMAGE=scratch
+# distroless images are signed by cosign. You should verify the image with the following public key:
```
I confirm I could verify locally :+1:.

qmonnet commented on Documentation/operations/upgrade.rst:
```diff
@@ -398,6 +398,8 @@ Helm Options
   If you were leveraging ``securityContext.extraCapabilities``, you need to review
   ``securityContext.capabilities.cilium_agent``.
+* The ``securityContext`` for Hubble Relay now applies to the container, not the pod. To update the security context of the pod, use ``podSecurityContext``.
+* The ``securityContext`` for Hubble Relay now defaults to drop all capabilities and run as non-root user.
```
I think you'll need a new section for v1.14 for those? This PR is not making it to v1.13 as far as I understand.

rolinh replied:
Oh right, good point!

rolinh replied:
Should I go ahead and remove the v1.13 section?

qmonnet replied:
I believe maintainers do it as part as the preparation steps for the next release? [EDIT: Joe did last time, but maybe seems to be an isolated commit, see 6329429c6cc821c9f4dff6f7f19cf5a8f876c81a] I don't remember. Might be worth checking if there was any commit touching that section after we branched off from 1.13 in any case

qmonnet replied:
```
$ git log --oneline $(git merge-base master v1.13).. -- Documentation/operations/upgrade.rst
51063421d8a2 metrics: Remove unused ARP ping metrics
6be650b98007 docs: Bump Linux minimum version to 4.19.57
dcc458b5f423 hubble: Update the reason label for hubble_drop_total metric
50ed9386696a Make cilium pprof listen address configurable
a0eeb41b19fd Update Documentation/operations/upgrade.rst
fb98a03314bc Add upgrade note for hubble_policy_verdicts_total now listing L7 flows
5b9188b86456 k8s/watchers: Adjust CNP handlers to set ProcessStartTime for PolicyAdd
fda366730e4d docs: Document internalTrafficPolicy-related changes in the upgrade guide
```

Some of these may be worth reporting to the new section, if they haven't been backported to 1.13.

rolinh replied:
I pushed a commit to remove the v1.13 upgrade notes section. I'll double check that upgrade notes for v1.14 weren't added to the v1.13 section.

rolinh replied:
All right, all of the commits in your message above have been backported to v1.13 so I think we are good to go.

rolinh committed:
docs: remove 1.13 and earlier upgrade docs

These docs are already stored in the v1.12 branch and exposed via the
corresponding docs on docs.cilium.io, so there is no need to include
that text directly in the page. Removing it will also help developers to
update the correct notes for the upcoming release.

Signed-off-by: Robin Hahling <robin.hahling@gw-computing.net>

qmonnet reviewed (approved):
Looks good to me, nice!

sayboras reviewed (approved):
LGTM :heavy_check_mark:

kaworu reviewed (approved):
:rocket:
